File: sdks/dotnet/src/Evaluator.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using PhaseFlag.Models;

namespace PhaseFlag;

/// <summary>
/// Local feature flag evaluation engine.
/// <para>
/// Evaluates a <see cref="FlagDefinition"/> against an <see cref="EvaluationContext"/>
/// using targeting rules, condition matching, and deterministic percentage rollouts
/// (DJB2 hashing). Rules are evaluated in ascending priority order; the first match wins.
/// </para>
/// </summary>
public static class Evaluator
{
    // -----------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------

    /// <summary>
    /// Evaluate a flag definition locally against the given context.
    /// </summary>
    /// <param name="flag">The flag definition to evaluate.</param>
    /// <param name="ctx">
    /// Optional evaluation context. When <c>null</c>, an empty context is used.
    /// </param>
    /// <returns>
    /// An <see cref="EvaluationResult"/> with the resolved variation, or <c>null</c>
    /// if the flag has no variations at all.
    /// </returns>
    public static EvaluationResult? Evaluate(FlagDefinition flag, EvaluationContext? ctx)
    {
        ctx ??= new EvaluationContext();

        var variationsById = new Dictionary<string, Variation>(flag.Variations.Count);
        foreach (var v in flag.Variations)
        {
            variationsById[v.Id] = v;
        }

        EvaluationResult MakeResult(string variationId, string reason)
        {
            variationsById.TryGetValue(variationId, out var variation);
            return new EvaluationResult
            {
                FlagKey = flag.Key,
                VariationId = variationId,
                VariationKey = variation?.Key,
                Value = variation?.Value ?? default,
                Reason = reason,
            };
        }

        // Sort targeting rules by priority (ascending -- lower number = higher priority)
        var rules = flag.TargetingRules
            .OrderBy(r => r.Priority)
            .ToList();

        foreach (var rule in rules)
        {
            if (!MatchAllConditions(rule.Conditions, ctx))
                continue;

            // 1. Explicit variation
            if (!string.IsNullOrEmpty(rule.VariationId))
            {
                return MakeResult(rule.VariationId, "targeting_match");
            }

            // 2. Percentage rollout
            if (rule.PercentageRollout is not null)
            {
                var userId = ctx.UserId ?? ctx.SessionId ?? string.Empty;
                var resolvedId = ResolvePercentageRollout(rule.PercentageRollout, flag.Key, userId);
                if (resolvedId is not null)
                {
                    return MakeResult(resolvedId, "percentage_rollout");
                }
            }
        }

        // No rule matched -- return default variation
        return MakeResult(flag.DefaultVariationId, "default");
    }

    // -----------------------------------------------------------------
    // Condition matching
    // -----------------------------------------------------------------

    /// <summary>
    /// Evaluate a single targeting condition against an evaluation context.
    /// </summary>
    /// <param name="cond">The condition to evaluate.</param>
    /// <param name="ctx">
    /// Optional evaluation context. When <c>null</c>, only negation operators match.
    /// </param>
    /// <returns><c>true</c> if the condition is satisfied.</returns>
    public static bool MatchCondition(TargetingCondition cond, EvaluationContext? ctx)
    {
        var actual = ctx?.Get(cond.Attribute);
        var op = cond.Operator;
        var target = cond.Value;

        // Missing attribute never matches (except negation operators)
        if (actual is null)
        {
            return op is "is_not" or "not_contains" or "not_one_of";
        }

        var actualStr = ConvertToString(actual);

        return op switch
        {
            "is" => actualStr == JsonElementToString(target),
            "is_not" => actualStr != JsonElementToString(target),
            "contains" => actualStr.Contains(JsonElementToString(target), StringComparison.Ordinal),
            "not_contains" => !actualStr.Contains(JsonElementToString(target), StringComparison.Ordinal),
            "one_of" => MatchOneOf(actualStr, target),
            "not_one_of" => !MatchOneOf(actualStr, target),
            "gt" => CompareNumeric(actualStr, target, gt: true),
            "lt" => CompareNumeric(actualStr, target, gt: false),
            "matches_regex" => MatchRegex(actualStr, target),
            _ => false,
        };
    }

    // -----------------------------------------------------------------
    // DJB2 hashing
    // -----------------------------------------------------------------

    /// <summary>
    /// Compute a DJB2 hash of the input string, returning an unsigned 32-bit integer.
    /// <para>
    /// The algorithm seeds at 5381 and for each character computes
    /// <c>hash = ((hash &lt;&lt; 5) + hash) + char</c>.
    /// </para>
    /// </summary>
    /// <param name="input">The string to hash.</param>
    /// <returns>An unsigned 32-bit hash value.</returns>
    public static uint Djb2Hash(string input)
    {
        uint hash = 5381;
        foreach (var ch in input)
        {
            hash = unchecked(((hash << 5) + hash) + (uint)ch);
        }
        return hash;
    }

    /// <summary>
    /// Determine the variation ID via percentage rollout using DJB2 hashing.
    /// </summary>
    /// <param name="rollout">The percentage rollout configuration.</param>
    /// <param name="flagKey">The flag key, combined with <paramref name="userId"/> for hashing.</param>
    /// <param name="userId">The user identifier for deterministic bucketing.</param>
    /// <returns>
    /// The resolved variation ID, or <c>null</c> if no user ID is available
    /// or the bucket falls outside all slices.
    /// </returns>
    public static string? ResolvePercentageRollout(PercentageRollout rollout, string flagKey, string userId)
    {
        if (string.IsNullOrEmpty(userId))
            return null;

        var bucket = Djb2Hash($"{flagKey}:{userId}") % 100;
        uint cumulative = 0;

        foreach (var entry in rollout.Variations)
        {
            cumulative += (uint)entry.Weight;
            if (bucket < cumulative)
            {
                return entry.VariationId;
            }
        }

        return null;
    }

    // -----------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------

    private static bool MatchAllConditions(List<TargetingCondition> conditions, EvaluationContext ctx)
    {
        // All conditions must match (AND logic)
        foreach (var condition in conditions)
        {
            if (!MatchCondition(condition, ctx))
                return false;
        }
        return true;
    }

    /// <summary>
    /// Convert an arbitrary object to its string representation.
    /// </summary>
    private static string ConvertToString(object? value)
    {
        return value?.ToString() ?? string.Empty;
    }

    /// <summary>
    /// Convert a <see cref="JsonElement"/> to a plain string for comparison.
    /// </summary>
    private static string JsonElementToString(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.String => element.GetString() ?? string.Empty,
            JsonValueKind.Number => element.GetRawText(),
            JsonValueKind.True => "True",
            JsonValueKind.False => "False",
            JsonValueKind.Null => string.Empty,
            _ => element.GetRawText(),
        };
    }

    /// <summary>
    /// Check whether <paramref name="actual"/> is contained in the target list.
    /// The target may be a JSON array or a single value.
    /// </summary>
    private static bool MatchOneOf(string actual, JsonElement target)
    {
        if (target.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in target.EnumerateArray())
            {
                if (actual == JsonElementToString(item))
                    return true;
            }
            return false;
        }

        // Single value fallback
        return actual == JsonElementToString(target);
    }

    /// <summary>
    /// Numeric comparison. When <paramref name="gt"/> is true, checks actual &gt; target;
    /// otherwise checks actual &lt; target.
    /// </summary>
    private static bool CompareNumeric(string actual, JsonElement target, bool gt)
    {
        if (!double.TryParse(actual, System.Globalization.NumberStyles.Float,
                System.Globalization.CultureInfo.InvariantCulture, out var a))
            return false;

        double b;
        if (target.ValueKind == JsonValueKind.Number)
        {
            b = target.GetDouble();
        }
        else
        {
            var targetStr = JsonElementToString(target);
            if (!double.TryParse(targetStr, System.Globalization.NumberStyles.Float,
                    System.Globalization.CultureInfo.InvariantCulture, out b))
                return false;
        }

        return gt ? a > b : a < b;
    }

    /// <summary>
    /// Check whether <paramref name="actual"/> matches the regex pattern in <paramref name="target"/>.
    /// </summary>
    private static bool MatchRegex(string actual, JsonElement target)
    {
        try
        {
            var pattern = JsonElementToString(target);
            return Regex.IsMatch(actual, pattern);
        }
        catch (RegexParseException)
        {
            return false;
        }
    }
}

File: sdks/php/src/Evaluator.php
<?php

declare(strict_types=1);

namespace PhaseFlag;

use PhaseFlag\Models\EvaluationContext;
use PhaseFlag\Models\EvaluationResult;
use PhaseFlag\Models\FlagDefinition;
use PhaseFlag\Models\PercentageRollout;
use PhaseFlag\Models\TargetingCondition;
use PhaseFlag\Models\Variation;

/**
 * Local feature flag evaluation engine.
 *
 * Mirrors the server-side evaluation logic so that flags can be resolved
 * entirely in-process without a network round-trip. Supports targeting
 * conditions (9 operators), percentage rollouts using DJB2 hashing, and
 * priority-ordered rule matching.
 */
final class Evaluator
{
    /**
     * Evaluate a flag definition against an evaluation context.
     *
     * Rules are checked in priority order (ascending). The first rule whose
     * conditions all match determines the result. If no rule matches the
     * default variation is returned.
     */
    public static function evaluate(FlagDefinition $flag, ?EvaluationContext $ctx): EvaluationResult
    {
        $ctx ??= new EvaluationContext();

        /** @var array<string, Variation> $variationsById */
        $variationsById = [];
        foreach ($flag->variations as $variation) {
            $variationsById[$variation->id] = $variation;
        }

        $makeResult = static function (string $variationId, string $reason) use ($flag, $variationsById): EvaluationResult {
            $v = $variationsById[$variationId] ?? null;
            return new EvaluationResult(
                flagKey: $flag->key,
                variationId: $variationId,
                variationKey: $v?->key,
                value: $v?->value,
                reason: $reason,
            );
        };

        // Sort targeting rules by priority (lower number = higher priority).
        $rules = $flag->targetingRules;
        usort($rules, static fn ($a, $b) => $a->priority <=> $b->priority);

        foreach ($rules as $rule) {
            // All conditions must match (AND logic).
            $allMatch = true;
            foreach ($rule->conditions as $condition) {
                if (!self::matchCondition($condition, $ctx)) {
                    $allMatch = false;
                    break;
                }
            }

            if (!$allMatch) {
                continue;
            }

            // 1. Explicit variation.
            if ($rule->variationId !== null) {
                return $makeResult($rule->variationId, 'targeting_match');
            }

            // 2. Percentage rollout.
            if ($rule->percentageRollout !== null) {
                $userId = $ctx->userId ?? $ctx->sessionId ?? '';
                $vid = self::resolvePercentageRollout($rule->percentageRollout, $flag->key, $userId);
                if ($vid !== null) {
                    return $makeResult($vid, 'percentage_rollout');
                }
            }
        }

        // No rule matched -- return the default variation.
        return $makeResult($flag->defaultVariationId, 'default');
    }

    /**
     * Evaluate a single targeting condition against an evaluation context.
     *
     * Supported operators:
     *  - is, is_not
     *  - contains, not_contains
     *  - one_of, not_one_of
     *  - gt, lt
     *  - matches_regex
     */
    public static function matchCondition(TargetingCondition $cond, ?EvaluationContext $ctx): bool
    {
        $ctx ??= new EvaluationContext();
        $actual = $ctx->get($cond->attribute);
        $operator = $cond->operator;
        $target = $cond->value;

        // Missing attribute never matches except negation operators.
        if ($actual === null) {
            return in_array($operator, ['is_not', 'not_contains', 'not_one_of'], true);
        }

        return match ($operator) {
            'is'            => (string) $actual === (string) $target,
            'is_not'        => (string) $actual !== (string) $target,
            'contains'      => str_contains((string) $actual, (string) $target),
            'not_contains'  => !str_contains((string) $actual, (string) $target),
            'one_of'        => self::matchOneOf($actual, $target),
            'not_one_of'    => !self::matchOneOf($actual, $target),
            'gt'            => self::compareNumeric($actual, $target, '>'),
            'lt'            => self::compareNumeric($actual, $target, '<'),
            'matches_regex' => self::matchRegex($actual, $target),
            default         => false,
        };
    }

    /**
     * DJB2 string hash returning an unsigned 32-bit integer.
     *
     * Mirrors the server-side implementation used for deterministic
     * percentage rollout bucketing.
     */
    public static function djb2Hash(string $input): int
    {
        $hash = 5381;

        for ($i = 0, $len = strlen($input); $i < $len; $i++) {
            $hash = (($hash << 5) + $hash + ord($input[$i])) & 0xFFFFFFFF;
        }

        return $hash;
    }

    /**
     * Determine the variation ID via percentage rollout using DJB2 hashing.
     *
     * Returns null if the user ID is empty or no bucket matches.
     */
    public static function resolvePercentageRollout(
        PercentageRollout $rollout,
        string $flagKey,
        string $userId,
    ): ?string {
        if ($userId === '') {
            return null;
        }

        $bucket = self::djb2Hash("{$flagKey}:{$userId}") % 100;

        $cumulative = 0;
        foreach ($rollout->variations as $entry) {
            $cumulative += $entry->weight;
            if ($bucket < $cumulative) {
                return $entry->variationId;
            }
        }

        return null;
    }

    // ------------------------------------------------------------------
    // Private helpers
    // ------------------------------------------------------------------

    /**
     * Check whether the actual value is contained in the target list.
     */
    private static function matchOneOf(mixed $actual, mixed $target): bool
    {
        if (is_array($target)) {
            return in_array((string) $actual, array_map('strval', $target), true);
        }

        return (string) $actual === (string) $target;
    }

    /**
     * Numeric comparison.  Returns false if either side is not numeric.
     */
    private static function compareNumeric(mixed $actual, mixed $target, string $op): bool
    {
        if (!is_numeric($actual) || !is_numeric($target)) {
            return false;
        }

        $a = (float) $actual;
        $b = (float) $target;

        return match ($op) {
            '>'     => $a > $b,
            '<'     => $a < $b,
            default => false,
        };
    }

    /**
     * Regex match with error suppression for invalid patterns.
     */
    private static function matchRegex(mixed $actual, mixed $target): bool
    {
        $pattern = '/' . str_replace('/', '\\/', (string) $target) . '/';

        // Suppress warnings from malformed patterns.
        $result = @preg_match($pattern, (string) $actual);

        return $result === 1;
    }
}
